src/NuGet.Protocol.Core.v2/DownloadResourceV2.cs

```diff
@@ -66,5 +69,211 @@ public DownloadResourceV2(V2Resource resource)
+                    {
+                        newPackage = localPackage;
+                    }
+                }
+                catch
+                {
+                    // Ignore cache failures here to match NuGet.Core
+                    // The bad package will be deleted and replaced during the download.
+                }
+
+                // If the local package does not exist in the cache download it from the source
+                if (newPackage == null)
+                {
+                    newPackage = DownloadToMachineCache(
+                        cacheRepository,
+                        package,
+                        dataServiceRepo,
+                        package.DownloadUri,
+                        token);
+                }
+
+                // Read the package from the machine cache
+                if (newPackage != null)
+                {
+                    return new DownloadResourceResult(newPackage.GetStream());
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new NuGetProtocolException(Strings.FormatProtocol_FailedToDownloadPackage(
+                    package,
+                    dataServiceRepo.Source),
+                    ex);
+            }
+
+            return null;
+        }
+
+        private static DownloadResourceResult DownloadFromIdentity(
+            PackageIdentity identity, 
+            IPackageRepository repository, 
+            CancellationToken token)
+        {
+            var version = SemanticVersion.Parse(identity.Version.ToString());
+            var dataServiceRepo = repository as DataServicePackageRepository;
+
+            if (dataServiceRepo != null)
+            {
+                // Clone the repo to allow for concurrent calls
+                var sourceUri = new Uri(dataServiceRepo.Source);
+                dataServiceRepo = new DataServicePackageRepository(sourceUri);
+
+                var package = dataServiceRepo.FindPackage(identity.Id, version);
+                token.ThrowIfCancellationRequested();
+
+                // For online sources get the url and retrieve it with cancel support
+                var dataServicePackage = package as DataServicePackage;
+                var url = dataServicePackage.DownloadUrl;
+
+                var downloadedPackage = DownloadToMachineCache(
+                    MachineCache.Default,
+                    identity,
+                    dataServiceRepo,
+                    url,
+                    token);
+
+                if (downloadedPackage != null)
+                {
+                    return new DownloadResourceResult(downloadedPackage.GetStream());
+                }
+            }
+            else
+            {
+                var package = repository.FindPackage(identity.Id, version);
+
+                // Use a folder reader for unzipped repos
+                if (repository is UnzippedPackageRepository)
```

Comment:
Does this work with both

1. Flat nupkg folder
2. V3 nupkg structure (not unzipped)